Include/internal/pycore_critical_section.h

```diff
@@ -109,8 +109,18 @@ _PyCriticalSection_IsActive(uintptr_t tag)
 static inline void
 _PyCriticalSection_BeginMutex(PyCriticalSection *c, PyMutex *m)
 {
+    // As an optimisation for locking the same object recursively, skip
+    // locking if the mutex is currently locked and the topmost, active,
+    // single critical section.
+    PyThreadState *tstate = _PyThreadState_GET();
+    if (tstate->critical_section &&
+        !(tstate->critical_section & _Py_CRITICAL_SECTION_MASK) &&
+        ((PyCriticalSection *)tstate->critical_section)->_cs_mutex == m) {
+        c->_cs_mutex = NULL;
+        c->_cs_prev = 0;
+        return;
+    }
```

Comment:
I think we should push this down into the non-inlined function `_PyCriticalSection_BeginSlow()`. I'd like to keep the inlined bit small and this is a relatively rare case compared to the common fast path lock acquisition.

I also don't think we need the `!(tstate->critical_section & _Py_CRITICAL_SECTION_MASK)` check. The top most critical section is always active if you have a valid thread state. It's also ok if it's a PyCriticalSection2 (and we only check the first mutex). Every `PyCriticalSection2` is also a  `PyCriticalSection`.

Reply:
Moving it to the start of the slow path makes sense. Done. (Benchmarks pending, but I imagine it won't matter much for the things that the change improved.)

I wasn't sure if there was a path where the topmost critical section would be inactive, but I guess that can only happen if the current thread isn't active, in which case no new critical sections should be possible either. I switched to just using untag_critical_section now that the code is in `_BeginSlow`.